### PR TITLE
Coin v6 upgrade

### PR DESCRIPTION
**Pull request by edmundnoble**
This PR is meant to be reviewed commit-by-commit.

This PR includes Coin v6, intended to prevent rotation of principals away from their corresponding guards. See the KIP for more details while it's pending merge at https://github.com/kadena-io/KIPs/pull/50.

There are tests for this in the new Coin v6 REPL test, as well as the new upgrade test for Chainweb 2.23.

The intended changes in this PR are:
- Coin v6 is added, mostly identical to Coin v5 with the change that rotate no longer allows rotating principal accounts away from their corresponding guards; a principal which already has a mismatched guard is allowed to rotate back to its corresponding guard.
- Coin v6 blesses Coin v5's hash (and all previous Coin hashes) to allow continued operation of existing code referencing Coin.
- Fast-devnet is updated to include Coin v6 at genesis.
- Mainnet, testnet, and all testing versions with Pact enabled are updated to include Coin v6 at the Chainweb 2.23 fork.
- Testing versions now enable the Chainweb 2.23 fork after all previous forks.

### CODE REVIEW

File: pact/coin-contract/v6/coin-v6.pact [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+
+          (update allocation-table account
+            { "redeemed" : true
+            , "balance" : 0.0
+            })
+
+          "Allocation successfully released to main ledger"))
+    )))
+
+)
```
Comment (emilypi) [inline pact/coin-contract/v6/coin-v6.pact:690]: FYI for reviewers, here's the diff: 

```pact
31a32,34
>   ;; coin v5
>   (bless "rE7DU8jlQL9x_MPYuniZJf5ICBTAEHAIFQCB4blofP4")
> 
315a319,325
> 
>       ; Allow rotation only for vanity accounts, or
>       ; re-rotating a principal account back to its proper guard
>       (enforce (or (not (is-principal account))
>                   (validate-principal new-guard account))
>         "It is unsafe for principal accounts to rotate their guard")
> 
318d327
< 
320d328
< 
```

File: pact/coin-contract/v6/coin-v6.repl [hunk 2]
```diff
@@ -0,0 +1,1174 @@
...
+
+(expect "rotate works for same k:account with the corresponding principal guard"
+  "Write succeeded"
+  (coin.rotate "k:5b4c9fc5207fcf700a5fbcb48c261820149c8ecd52e005282ed9e3f5febcd1f4" (read-keyset "k")))
+
+(expect "coin rotation works for vanity accounts"
+  "Write succeeded"
+  (coin.rotate "emily" (read-keyset "k1")))
+
+(rollback-tx)
```
Comment (emilypi) [inline pact/coin-contract/v6/coin-v6.repl:1174]: FYI for reviewers: 

```pact
2d1
< (env-exec-config ["DisableInlineMemCheck", "DisablePact43", "DisablePact44", "DisablePact45"])
12c11
<   840 (env-gas))
---
>   60999 (env-gas))
15c14,15
< (env-exec-config ["DisablePact44", "DisablePact45"])
---
> ; we disable pact 44 to allow using un-namespaced keysets
> (env-exec-config ["DisablePact44"])
26c26
< (load "coin-v5.pact")
---
> (load "coin-v6.pact")
30c30
<   66702
---
>   66828
67c67
<   225 (env-gas))
---
>   196 (env-gas))
412c412
<   572 (env-gas))
---
>   425 (env-gas))
449c449
<   "Gas cost of transfer-create" 576 (env-gas))
---
>   "Gas cost of transfer-create" 429 (env-gas))
911a912
> (commit-tx)
1131a1133
> 
1141a1144,1174
> 
> ; Test non-principal account rotating a guard is rejected
> (begin-tx)
> (env-data
>   {'k: ["5b4c9fc5207fcf700a5fbcb48c261820149c8ecd52e005282ed9e3f5febcd1f4"]
>   , "jose":["greg"]
>   , "k1":["keys1"]}
> 
>   )
> (env-sigs
>   [{ "key":"5b4c9fc5207fcf700a5fbcb48c261820149c8ecd52e005282ed9e3f5febcd1f4"
>    , "caps":[(coin.ROTATE "k:5b4c9fc5207fcf700a5fbcb48c261820149c8ecd52e005282ed9e3f5febcd1f4")]}
>    ,{"key":"keys1"
>    , "caps":[(coin.ROTATE "emily")]}
>    ,{"key":"keys2"
>    , "caps":[(coin.ROTATE "emily")]}])
> 
> (coin.create-account "k:5b4c9fc5207fcf700a5fbcb48c261820149c8ecd52e005282ed9e3f5febcd1f4" (read-keyset "k"))
> 
> (expect-failure "rotate fails for k: account"
>   (coin.rotate "k:5b4c9fc5207fcf700a5fbcb48c261820149c8ecd52e005282ed9e3f5febcd1f4" (read-keyset "jose")))
> 
> (expect "rotate works for same k:account with the corresponding principal guard"
>   "Write succeeded"
>   (coin.rotate "k:5b4c9fc5207fcf700a5fbcb48c261820149c8ecd52e005282ed9e3f5febcd1f4" (read-keyset "k")))
> 
> (expect "coin rotation works for vanity accounts"
>   "Write succeeded"
>   (coin.rotate "emily" (read-keyset "k1")))
> 
> (rollback-tx)
```

File: tools/ea/Ea/Genesis.hs [hunk 3]
```diff
@@ -143,8 +144,11 @@ coinContractV4 = "pact/coin-contract/v4/load-coin-contract-v4.yaml"
 coinContractV5 :: FilePath
 coinContractV5 = "pact/coin-contract/v5/load-coin-contract-v5.yaml"
 
-installCoinContractV5 :: FilePath
-installCoinContractV5 = "pact/coin-contract/v5/install-coin-contract-v5.yaml"
+coinContractV6 :: FilePath
+coinContractV6 = "pact/coin-contract/v6/load-coin-contract-v6.yaml"
+
+installCoinContractV6 :: FilePath
+installCoinContractV6 = "pact/coin-contract/v6/install-coin-contract-v6.yaml"
```
Comment (emilypi) [inline tools/ea/Ea/Genesis.hs:151]: what's the point of this "install" copy? I actually don't remember this being a thing, but it seems to be redundant in both this case and the previous v5 case.

**Review by emilypi — COMMENTED** [review]
This looks good, but we need to cover the forking tests in the Haskell tests for pact

File: tools/ea/Ea/Genesis.hs (hunk 3 above)
Comment (edmundnoble) [reply tools/ea/Ea/Genesis.hs:151]: This is for fast-devnet.

**Review by emilypi — APPROVED** [review]

File: tools/ea/Ea/Genesis.hs (hunk 3 above)
Comment (emilypi) [reply tools/ea/Ea/Genesis.hs:151]: great, thanks

**Review by jmcardon — APPROVED** [review]

**Pull request merged**